### PR TITLE
feat: use opendal as the s3 sdk by default

### PR DESCRIPTION
**Pull request by hzxa21**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Non-s3 object store backend has been using opendal for a while. This PR switches the s3 sdk to opendal by default as well.

related: #14321

There is a behavior change for s3 retry:

Prior to this PR:
- S3 operation retry attempts = attempts specified in config section [[storage.object_store.retry]](https://github.com/risingwavelabs/risingwave/blob/1bb9fa31ebf1130921980439d8c7a1f411c80083/src/config/example.toml#L201) * [default retry attempts in aws sdk s3](https://docs.rs/aws-smithy-types/1.2.2/src/aws_smithy_types/retry.rs.html#321) (default = 3 * 2 = 6)

After this PR:
- S3 operation retry attempts = attempts specified in config section [[storage.object_store.retry]](https://github.com/risingwavelabs/risingwave/blob/1bb9fa31ebf1130921980439d8c7a1f411c80083/src/config/example.toml#L201) (default = 3)

The reason why we have this behavior change is because we didn't disable the internal retry when using aws-sdk-s3, which unexpectedly add more retries than specified in RW config. Given that all object stores other than s3 already honored the retry attempts specified in config `[storage.object_store.retry]`, I think it is okay to correct the unexpected behavior in S3 after switching to opendal as well and let it fully honor our config.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [x] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

The sdk that connects to S3 object store state backend is switched from [aws-sdk-s3](https://crates.io/crates/aws-sdk-s3) to [opendal](https://crates.io/crates/opendal). This is an internal 3rd party dependency change, intended to unify the sdk RisingWave uses to connect to different object store backend and reduce burden on maintenance. This change should be seamingless but if you experience instability or unexpected errors when running RisingWave on S3, please file an issue in our GitHub repo.

We will keep [aws-sdk-s3](https://crates.io/crates/aws-sdk-s3) as a fallback option before we fully deprecate it in 1-2 releases. You can use the following config to switch back to aws-sdk-s3:
```toml
[storage.object_store.s3.developer]
use_opendal = false
```

### CODE REVIEW

**Comment by hzxa21** [comment]
The following longevity test passed with similar resource usage, source throughput and no compaction lag using the same nightly image before and after switching to opendal:
- [nexmark all queries 10K](https://buildkite.com/risingwave-test/longevity-test/builds/1709)
- [nexmark q8,q9,q15,q16 300K ](https://buildkite.com/risingwave-test/longevity-test/builds/1740#01914477-6ed1-41b8-bfc7-ed5cdbd7a19b)
- [nexmark q17,q18,q19,q20 300K ](https://buildkite.com/risingwave-test/longevity-test/builds/1737)
- [nexmark q101 q102 q103 q104 q105 300K ](https://buildkite.com/risingwave-test/longevity-test/builds/1727)
- [nexmark q3 q4 q5 q7 300K ](https://buildkite.com/risingwave-test/longevity-test/builds/1714)

**Comment by hzxa21** [comment]
nexmark perf result:
image
|Nexmark Name                               |Category          |A                 |B                 |percentage_change      |
|-------------------------------------------|------------------|------------------|------------------|-----------------------|
|nexmark-q3-no-condition-blackhole          |nightly-20240801-A|737488.98         |780340.81         |-5.49                  |
|nexmark-q5-rewrite-blackhole               |nightly-20240801-A|178604.56         |185153.18         |-3.54                  |
|nexmark-q105-blackhole                     |nightly-20240801-A|248458.21         |255995.12         |-2.94                  |
|nexmark-q104-blackhole                     |nightly-20240801-A|385924.43         |397419.77         |-2.89                  |
|nexmark-q103-blackhole                     |nightly-20240801-A|538433.35         |552686.84         |-2.58                  |
|nexmark-q6-group-top1-blackhole-watermark  |nightly-20240801-A|218919.25         |223105.68         |-1.88                  |
|nexmark-q17-blackhole                      |nightly-20240801-A|557997.57         |565969.57         |-1.41                  |
|nexmark-q13-blackhole                      |nightly-20240801-A|902050.96         |912347.09         |-1.13                  |
|nexmark-q0-blackhole                       |nightly-20240801-A|961466.26         |971879.04         |-1.07                  |
|nexmark-q5-many-windows-blackhole-watermark|nightly-20240801-A|18659.28          |18801.35          |-0.76                  |
|nexmark-q18-blackhole-watermark            |nightly-20240801-A|274057.19         |274919.90         |-0.31                  |
|nexmark-q7-rewrite-blackhole               |nightly-20240801-A|944367.48         |943635.39         |0.08                   |
|nexmark-q19-blackhole                      |nightly-20240801-A|215850.20         |215533.11         |0.15                   |
|nexmark-q18-blackhole                      |nightly-20240801-A|302847.91         |301685.38         |0.39                   |
|nexmark-q101-blackhole                     |nightly-20240801-A|372370.46         |369920.53         |0.66                   |
|nexmark-q12-blackhole                      |nightly-20240801-A|860672.42         |851876.69         |1.03                   |
|nexmark-q8-blackhole                       |nightly-20240801-A|650701.28         |642525.48         |1.27                   |
|nexmark-q16-blackhole                      |nightly-20240801-A|108489.16         |106618.43         |1.75                   |
|nexmark-q9-blackhole-watermark             |nightly-20240801-A|221857.52         |216993.08         |2.24                   |
|nexmark-q20-blackhole                      |nightly-20240801-A|394281.68         |385083.45         |2.39                   |
|nexmark-q102-blackhole                     |nightly-20240801-A|233524.33         |227996.14         |2.42                   |
|nexmark-q5-blackhole-watermark             |nightly-20240801-A|359371.41         |348187.12         |3.21                   |
|nexmark-q8-blackhole-watermark             |nightly-20240801-A|628568.36         |605702.11         |3.78                   |
|nexmark-q5-blackhole                       |nightly-20240801-A|141079.84         |135382.16         |4.21                   |
|nexmark-q15-blackhole                      |nightly-20240801-A|481344.28         |460756.91         |4.47                   |
|nexmark-q7-blackhole                       |nightly-20240801-A|671364.28         |642561.12         |4.48                   |
|nexmark-q7-blackhole-watermark             |nightly-20240801-A|630618.80         |599397.81         |5.21                   |
|nexmark-q9-blackhole                       |nightly-20240801-A|232522.26         |214654.96         |8.32                   |
|nexmark-q4-blackhole                       |nightly-20240801-A|307474.27         |266900.33         |15.20                  |

**Review by wcy-fdu — APPROVED** [review]
Thanks for the effort in testing!

BTW, Should we make your testing workload as standard release testing, and do such tests every time bumping OpenDAL version and releasing RisingWave? Can we  configure some automatic jobs to complete one-click testing?

**Comment by hzxa21** [comment]
> BTW, Should we make your testing workload as standard release testing, and do such tests every time bumping OpenDAL version and releasing RisingWave?

All the testing I have done is from the regular testing pipeline:
- The 10K longevity test will be run daily and the 300K longevity test will be run weekly.
- Perf test is the standard nexmark perf test and it will be run daily.

> Can we configure some automatic jobs to complete one-click testing?

Can we trigger perf/longevity test in PR? (not an urgent request) cc @huangjw806

**Review by Li0k — APPROVED** [review]
LGTM and thanks for your great contribution.

I have some questions about this PR and would like to have some short descriptions so that we can easily trace the PR in the future!

1. when we officially switch to opendal, some behaviours will change, such as whether opendal retries streaming requests, e.g. streaming_read / streaming_upload (I'm not entirely sure about this detail, if there is a change, please add a short description)
2. after switching to opendal, the number of retries provided by the configuration may change from the previous one
- For example, when we use aws-sdk, by default we use the 2 retries provided by sdk, which is tricky, but after switching to opendal, the number of retries will be reduced.

**Comment by Li0k** [comment]
Test error: It seems that we are printing a lot of duplicate logs

image

**Comment by hzxa21** [comment]
> 1. when we officially switch to opendal, some behaviours will change, such as whether opendal retries streaming requests, e.g. streaming_read / streaming_upload (I'm not entirely sure about this detail, if there is a change, please add a short description)

Both streaming_read / streaming_upload retries are implemented for opendal object store so there is no change in this part.

> 2. after switching to opendal, the number of retries provided by the configuration may change from the previous one
> 
> * For example, when we use aws-sdk, by default we use the 2 retries provided by sdk, which is tricky, but after switching to opendal, the number of retries will be reduced.
> 

The retry attempts do change after this PR for S3. I document the behavior change in the PR description.

> 3. opendal not support `retry_unknown_service_error`, We have to find out if opendal's retry implementation contains all the error types/error codes we need.

OpenDAL doesn't expose the exact error codes so there is no way we can check error codes in our codes. However, OpenDAL will use the `ErrorKind::Unexpected` to represent errors it cannot recognize. Therefore, we can use that to implement `retry_unknown_service_error`. I have push a commit to support that: https://github.com/risingwavelabs/risingwave/pull/18011/commits/49be5104d7eb2f0bdad5cd96a433cb80a2f15835


@Li0k PTAL. 
These are all valid points. Thanks for the comments.

File: src/object_store/src/object/mod.rs [hunk 1]
```diff
@@ -567,6 +567,7 @@ impl<OS: ObjectStore> MonitoredObjectStore<OS> {
     pub async fn upload(&self, path: &str, obj: Bytes) -> ObjectResult<()> {
         let operation_type = OperationType::Upload;
         let operation_type_str = operation_type.as_str();
+        let media_type = self.media_type();
```
Comment (Li0k) [inline src/object_store/src/object/mod.rs:570]: nits: How about using the same variable name? media_type and engine_type are too similar.

**Review by Li0k — APPROVED** [review]
Rest LGTM, Thanks for the test with all the contributions .

**Pull request merged**